### PR TITLE
[BEAM-2255] implement ORDER BY

### PR DESCRIPTION
**Pull request by xumingming**

### CODE REVIEW

**Comment by xumingming** [comment]
R: @XuMingmin

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamSortRel.java [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    }
+
+    @Override public int compare(BeamSQLRow row1, BeamSQLRow row2) {
+      for (int i = 0; i < fieldsIndices.size(); i++) {
+        int fieldIndex = fieldsIndices.get(i);
+        int fieldRet = 0;
+        SqlTypeName fieldType = row1.getDataType().getFieldsType().get(fieldIndex);
+        if (row1.isNull(fieldIndex) && row2.isNull(fieldIndex)) {
+          continue;
+        } else if (row1.isNull(fieldIndex) && !row2.isNull(fieldIndex)) {
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamSortRel.java]: Good catch! There is actually dedicated syntax to specify this: `NULL FIRST`, `NULLS LAST`.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamSortRel.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    PCollection<BeamSQLRow> upstream = planCreator.popUpstream();
+
+    BeamSQLRowComparator comparator = new BeamSQLRowComparator(fieldIndices, orientation);
+    // first find the top (offset + count)
+    PCollection<BeamSQLRow> orderedStream =
+        upstream.apply("extractTopOffset_plus_Fetch",
+            Top.of(startIndex + count, comparator))
+        .apply("extractTopOffset_plus_Fetch__Flatten", Flatten.<BeamSQLRow>iterables());
+
+    if (startIndex > 0) {
```
Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamSortRel.java]: Yeah, a `subList` is sufficient.

**Comment by xumingming** [comment]
For the `unbounded` source question, `Top` not only support `GlobalWindow` but also other window strategy[1].

For the `Window` operator, you mean [2] right? To be honest I don't know SQL support such syntax before :( , after some reading I think `ORDER BY` has little to do with `Window`?

suppose we have the following SQL:

```sql
select id, name, count(id) over x 
from person 
window x as (PARTITION BY id) 
order by id desc, name asc nulls first 
limit 1 offset 1
```

the corresponding relational algebra tree is:

```bash
LogicalSort(sort0=[$0], sort1=[$1], dir0=[DESC], dir1=[ASC-nulls-first], offset=[1], fetch=[1])
  LogicalProject(id=[$0], name=[$1], EXPR$2=[COUNT($0) OVER (PARTITION BY $0 RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING)])
    EnumerableTableScan(table=[[person]])
```

The window functions(the `EXPR$2` above) is executed as part of the `LogicProject`, after that the LogicalSort(`ORDER BY LIMIT`) will be executed. i.e. `LogicalSort` take the whole `LogicalSort` as input, regardless of whether the `LogiccalProject` contains a window functions or not.

[1].https://github.com/xumingming/beam/blob/7e04924ee7b31e28326f761618173749a55789d0/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Top.java#L87-L87
[2].https://github.com/apache/calcite/blob/aed5bca7c5f49d117d7b794395ff4b6ad2d76bde/core/src/main/java/org/apache/calcite/rel/core/Window.java#L65-L65

**Comment by XuMingmin** [comment]
`WINDOW` in my previous comment refers to HOP/TUMBLE/SESSION windows added in BEAM-2006. 

With that window applied, I suppose the output of `Top` would be top N records for each window. For unbounded data that may be fair enough, but not sure when it's applied on a bounded source.

**Comment by xumingming** [comment]
According to the result we discussed offline， due to the semantic constraints of `WINDOW`s in SQL:

> WINDOW in SQL is actually `BEAM WINDOW` + `BEAM GroupBy`

it makes `ORDER BY` does not make much sense for WINDOWS other than GlobalWindow. So our `ORDER BY` will only support GlobalWindow, and will throw UnsupportedException is used with other window. And will update javadoc to describe this.

**Comment by xumingming** [comment]
Seems no easy way to check whether there is a WINDOW applied in the same level(WINDOW can be in current query, or nested query), will not throw the exception for now.

File: dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamSortRel.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+      RexLiteral offsetLiteral = (RexLiteral) offset;
+      startIndex = ((BigDecimal) offsetLiteral.getValue()).intValue();
+    }
+  }
+
+  @Override public PCollection<BeamSQLRow> buildBeamPipeline(
+      BeamPipelineCreator planCreator) throws Exception {
+    RelNode input = getInput();
+    PCollection<BeamSQLRow> upstream = BeamSQLRelUtils.getBeamRelInput(input)
+        .buildBeamPipeline(planCreator);
```
Comment (XuMingmin) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamSortRel.java]: Can check existing window here by `upstream.getWindowingStrategy()`, if it's not `GlobalWindow` an unsupport expcetion is thrown.

Comment (xumingming) [inline dsls/sql/src/main/java/org/apache/beam/dsls/sql/rel/BeamSortRel.java]: 👍 I was trying to check it from Calcite side..

**Comment by xumingming** [comment]
@jbonofre can you take a look at this one?

**Comment by jbonofre** [comment]
Absolutely !

R: @jbonofre

**Comment by jbonofre** [comment]
Merged,  can you please close this PR ? Thanks ! Much appreciated !

**Comment by xumingming** [comment]
Thanks @XuMingmin @jbonofre !

**Pull request closed without merging**